packages/kyc-oracle/KycOracle.js

```diff
@@ -0,0 +1,231 @@
+      } catch (err) {
+        console.log;
+        return res.status(500).send(err);
+      }
+    });
+
+    // Query for KYC status and update the whitelist
+    this.app.get("/status/:sessionId", async (req, res) => {
+      try {
+        const { data } = await axios.get("https://workflow-api.synaps.io/v2/session/info", {
+          headers: {
+            "Api-Key": this.apiKey,
+            "Session-Id": req.params.sessionId,
+          },
+        });
+
+        const validated = data.status === "VERIFIED";
+        const userAddress = await this.getAddressForSession(req.params.sessionId);
+
+        if (validated) {
+          const alreadyApproved = await this.whitelist.getApproval(userAddress);
+          if (!alreadyApproved) {
+            await this.updateGasEstimate("safeLow");
+            const gasEstimate = await this.whitelist.estimateGas.approveUser(userAddress, true);
+            this.whitelist.approveUser(userAddress, true, { gasLimit: gasEstimate, gasPrice: this.gasPrice });
+          }
+        }
+
+        return res.status(200).send(data);
+      } catch (err) {
+        console.log(err);
+        return res.status(500).send(err);
+      }
+    });
+
+    this.server = this.app.listen(port, () => {
+      console.log(`⭐️ KycOracle running on port ${this.server.address().port}`);
+    });
+  }
+
+  /**
+   * Initialises the Kyc Oracle so that it knows where to find the `Whitelist` contract
+   * @param  {string}  whitelistAddress The address of the current `Whitelist` contract
+   * @return {Promise}
+   */
+  async initialise(whitelistAddress) {
+    const network = await this.provider.getNetwork();
+    this.chainId = network.chainId;
+
+    this.whitelistContractDef = await this.loader.load({ contractName: "Whitelist" }, { abi: true, address: false });
+    this.whitelist = new ethers.Contract(whitelistAddress, this.whitelistContractDef.abi, this.wallet);
+
+    await this.updateGasEstimate("safeLow");
+    await this.createDB();
+  }
+
+  /**
+   * Update the gas estimate
+   * @param  {string}  Transaction speed (fastest, fast, safeLow)
+   * @return {Promise}
+   */
+  async updateGasEstimate(type) {
+    if (this.chainId === 100) {
+      this.gasPrice = ethers.utils.hexlify(1000000000);
+      return;
+    }
+
+    try {
+      // Get latest from ethGasStation
+      const { data } = await axios.get("https://ethgasstation.info/json/ethgasAPI.json");
+
+      if (data[type]) {
+        this.gasPrice = ethers.utils.hexlify((data[type] / 10) * 1e9);
+      } else {
+        this.gasPrice = ethers.utils.hexlify(20000000000);
+      }
+    } catch (err) {
+      console.log(`Error during gas estimation: ${err}`);
+      this.gasPrice = ethers.utils.hexlify(20000000000);
+    }
+  }
+
+  async createDB() {
+    const db = await sqlite.open({ filename: this.dbPath, driver: sqlite3.Database });
+    await db.run("CREATE TABLE IF NOT EXISTS users ( address text NOT NULL UNIQUE, session_id text NOT NULL )");
```

Comment:
Maybe put a uniqueness constraint on the pairs of address/session_id?

https://stackoverflow.com/questions/12763726/mysql-make-a-pair-of-values-unique

That would avoid the need for the ` < 1 ` checks later on.

Something like:

```
`CREATE TABLE IF NOT EXISTS users (
    address text NOT NULL UNIQUE,
    session_id text NOT NULL UNIQUE,
    UNIQUE (address, session_id)
)`
```

Reply:
I'm slightly uncomfortable with making `session_id` unique. We definitely hope it is, but we don't know how Synaps generate them.

I don't understand what you're suggesting here regarding the composite key. If address is `unique` (which it already is), by definition the combination of `address` and `session_id` in each row much be unique.